### PR TITLE
CUMULUS-3149 -- Update bulk deletion lambda to allow for concurrent deletion when CMR deletion option is selected 

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-3149](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3149)

### Fixed

- **CUMULUS-3149**
  - Updates the api `/graunles/bulkDelete` endpoit to take the
    follow configuration keys for the bulkDelete:
    - maxDbConnections - Defaults to concurrency , and generally should not be
        changed unltess troubleshooting performance concerns.
    - concurrency - Number of concurrent bulk deletions to process at a time.
        Defaults to 10, increasing this value may improve throughput at the cost
        of additional database/CMR/etc load.
  - Updates all bulk api endpoints to add knexDebug boolean query parameter to
    allow for debugging of database connection issues in the future.
  - Fixed logic defect in bulk deletion logic where an information query was
    nested in a transaction call, resulting in transactions holding knex
    connection pool connections in a blocking way that would not resolve,
    resulting in deletion failures.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: CHANGELOG.md
Comment (npauzenga) [inline CHANGELOG.md:14]:
```suggestion
    following configuration keys for the bulkDelete:
```

File: CHANGELOG.md
Comment (npauzenga) [inline CHANGELOG.md:19]:
```suggestion
    - concurrency - Number of concurrent bulk deletions to process at a time.
            Defaults to 10, increasing this value may improve throughput at the cost
            of additional database/CMR/etc load.
    - maxDbConnections - Defaults to `concurrency`, and generally should not be
        changed unltess troubleshooting performance concerns.
```

File: packages/api/endpoints/granules.js
Comment (npauzenga) [inline packages/api/endpoints/granules.js:555]: Should we explicitly check this? It's going to be a string I guess so if I submit a payload with `KNEX_DEBUG: 'False'` this would turn out truthy.

Maybe `toLowerCase(payload.knexDebug) === 'true' || false` to not accept any string value.

File: packages/api/endpoints/granules.js
Comment (npauzenga) [inline packages/api/endpoints/granules.js:588]: Should we check that it's a valid integer after casting it as a number?

File: packages/api/lambdas/bulk-operation.js
Comment (npauzenga) [inline packages/api/lambdas/bulk-operation.js:81]: Reminder to udpate the docstring

Comment (npauzenga) [reply packages/api/lambdas/bulk-operation.js:81]: Also is it weird that this could be undefined? I see that we're assigning a default in `unpublishGranule` and that's the only place it's used so maybe not 🤔

File: packages/api/lib/granule-remove-from-cmr.js
Comment (npauzenga) [inline packages/api/lib/granule-remove-from-cmr.js:54]: Docstring 😎

File: packages/api/lib/granule-remove-from-cmr.js
Comment (npauzenga) [inline packages/api/lib/granule-remove-from-cmr.js:51]: We should probably mark this as optional in the docstring if it is.

File: packages/api/tests/lambdas/test-bulk-granule-delete.js
Comment (npauzenga) [inline packages/api/tests/lambdas/test-bulk-granule-delete.js:130]: Could we add a comment here?

**Review by npauzenga — CHANGES_REQUESTED** [review]
Looks great @Jkovarik! Just a couple questions.

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:555]: Sure thing: 8e3b3f9635

Comment (Jkovarik) [reply packages/api/endpoints/granules.js]: 8e3b3f9635

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:588]: Hmm.    In reality we should probably be doing a full validation on the input payload.        For the moment, how about that/fixing the units:

213e8342c4

File: packages/api/lambdas/bulk-operation.js
Comment (Jkovarik) [reply packages/api/lambdas/bulk-operation.js:80]: Docstring updated: 60ebe10f96

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js:55]: f4bcf807de

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js:55]: That's fair.   Fixed docstring, including defaults.  c1835721b6

File: packages/api/tests/lambdas/test-bulk-granule-delete.js
Comment (Jkovarik) [reply packages/api/tests/lambdas/test-bulk-granule-delete.js]: Yes, but I think making it into a descriptive variable probably works as well/better - how's this: a4e5d9a9c5

**Comment by Jkovarik** [comment]
@npauzenga  I believe I've addressed the feedback on this PR.   Take a look and let me know what you think!

File: packages/api/tests/endpoints/granules/test-bulk-delete.js
Comment (npauzenga) [inline packages/api/tests/endpoints/granules/test-bulk-delete.js:339]: Nice.

**Review by npauzenga — APPROVED** [review]
Nice work @Jkovarik! There's one suggested change on the changelog remaining but otherwise 👍. Thanks for addressing the comments!

File: packages/api/endpoints/granules.js
Comment (marchuffnagle) [inline packages/api/endpoints/granules.js:555]: This will fail if `payload.knexDebug` is undefined. Are we guaranteeing that it'll be present somewhere?

File: packages/api/endpoints/granules.js
Comment (marchuffnagle) [inline packages/api/endpoints/granules.js:593]: If we're making sure that these are integers, why are we sending them through `Number()` above?

File: packages/api/endpoints/granules.js
Comment (marchuffnagle) [inline packages/api/endpoints/granules.js:667]: What is the type of `envVars`? I'd guess that it's probably env vars that should be set in some other process. If that's the case, then the type is probably `{ [k: string]: string }`. We're using `false` below, which should probably be `'false'` in that case.

Comment (Jkovarik) [reply packages/api/endpoints/granules.js:673]: Agreed - and particularly in the case of KNEX_DEBUG, as 

```

export const isKnexDebugEnabled = (
  env: NodeJS.ProcessEnv = {}
) => env.KNEX_DEBUG === 'true';
```

**Comment by Jkovarik** [comment]
@marchuffnagle appreciate the additions - better-zod-error's approach is great!     There were a couple of minor bugs in the package, I'm thinking we should probably whip up some basic unit coverage there,  however that clearly exceeds the scope of this ticket.   Thoughts?

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [inline packages/api/endpoints/granules.js:583]: @marchuffnagle  I feel like there should be a better way to organize these.     It's a net positive on this PR, but I think we should debate a structure if we're going to wholesale do this across the api package.

**Review by npauzenga — APPROVED** [review]
This looks great @Jkovarik you've addressed my comments and the zod addition looks awesome.

File: packages/api/endpoints/granules.js
Comment (marchuffnagle) [inline packages/api/endpoints/granules.js:602]: This should be noted as a breaking change since the API will be returning different error output now. Still worth doing, though.

File: packages/api/lib/granule-remove-from-cmr.js
Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:58]: Do we want to be calling this function here? If so, we need to provide 2 arguments.

File: packages/api/lib/granule-remove-from-cmr.js
Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:46]:
```suggestion
 * @param {Function} [params.removeGranuleFromCmrFunction=_removeGranuleFromCmr()]
```

Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:46]: What is the type of this function? What arguments should it take, and what should it return?

File: packages/api/lib/granule-remove-from-cmr.js
Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:39]:
```suggestion
 * @param {import('knex').Knex} params.knex - DB client
```

File: packages/api/lib/granule-remove-from-cmr.js
Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:41]: From the code, it looks like this is an object with name and version properties, not a string.

File: packages/api/lib/granule-remove-from-cmr.js
Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:44]: Probably not worth holding this PR up for, but we should have a more specific type for this.

File: packages/api/lib/granule-remove-from-cmr.js
Comment (marchuffnagle) [inline packages/api/lib/granule-remove-from-cmr.js:65]:
```suggestion
    /** @type {string} */
    let collectionId;
```

File: packages/api/lambdas/bulk-operation.js
Comment (marchuffnagle) [inline packages/api/lambdas/bulk-operation.js:75]: What is the type of this function?

File: packages/api/lambdas/bulk-operation.js
Comment (marchuffnagle) [inline packages/api/lambdas/bulk-operation.js:82]: `payload.concurrency` needs to be added to the JSDocs.

File: packages/api/lambdas/bulk-operation.js
Comment (marchuffnagle) [inline packages/api/lambdas/bulk-operation.js:83]: `payload.maxDbConnections` needs to be added to the JSDocs.

Comment (marchuffnagle) [inline packages/api/lambdas/bulk-operation.js:83]:
```suggestion
  process.env['dbMaxPool'] = payload.maxDbConnections || concurrency;
```

At some point we're going to enable the [noPropertyAccessFromIndexSignature](https://www.typescriptlang.org/tsconfig#noPropertyAccessFromIndexSignature) option, which will require this change. May as well make it now.

**Review by marchuffnagle — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (Jkovarik) [reply CHANGELOG.md:47]: I added a comment to the CL as it's a neutral change, and consistent with the rest of the CL entry, but •really• this should be documented here: https://github.com/nasa/cumulus-api/pull/331/files 

Good catch: 701657f22b

File: packages/common/src/index.ts
Comment (Jkovarik) [reply packages/common/src/index.ts]: 9c7eb97143

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:602]: Hmm, that's a good point, I had been thinking it returns the same status code, but the *text* return is changing. 

In that case..... I think I'm going to remove zod from this PR and reset the responses, as we don't want to be making breaking changes to the backport release series unless it's a critical bugfix.

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js]: Good catch: 19d5b71b5e

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js]: This (and lower docstring comments) updated 3ab40c4345  -- some of the TS includes *do* obnoxiously break our valid-jsdoc linter, but it's a warning and allows ts-check to properly infer typing.     We should talk about  a better approach for the repo/process/etc.

Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js]: 3ab40c4345

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js]: 3ab40c4345

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js]: Sure it is.  3ab40c4345.        To some degree we need to have a conversation on the bounds on ts-check related concerns/jsdoc typings/etc for consistency in review.       For example, the hack we're doing for this object with `cmr_link`.

File: packages/api/lib/granule-remove-from-cmr.js
Comment (Jkovarik) [reply packages/api/lib/granule-remove-from-cmr.js:66]: Still feel we should have a discussion re: the effort impact of this type annotation versus just doing it right in TS.        Obviously don't disagree with the intent. 

02b69e09bb

File: packages/api/lambdas/bulk-operation.js
Comment (Jkovarik) [reply packages/api/lambdas/bulk-operation.js]: I have slightly mixed feelings generally about how these unit mocks should be typed. 

 That said, it's typed to _removeGranuleFromCmr in the dependent function, which we could import (gross) to correctly type the method, or provide explicit typing.    I'm going with the latter, given TS inclinations, but open to feedback on this.        There's probably a refactor here as well, but that's solidly outside what I'm willing to scope for this ticket, absent strong objections from the team.

File: packages/api/lambdas/bulk-operation.js
Comment (Jkovarik) [reply packages/api/lambdas/bulk-operation.js:89]: e82fd4364f

File: packages/api/lambdas/bulk-operation.js
Comment (Jkovarik) [reply packages/api/lambdas/bulk-operation.js:88]: 951dbcd36f

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:602]: On second thought -- we could maintain error output via this sort of thing and avoid 'breaking' changes for the existing error check: 

2812e1f9d1

Opinions?

Comment (Jkovarik) [reply packages/api/endpoints/granules.js]: On second thought -- we could maintain error output via this sort of thing and avoid 'breaking' changes for the existing error check:

2812e1f9d1

Opinions?

File: packages/api/lambdas/bulk-operation.js
Comment (Jkovarik) [reply packages/api/lambdas/bulk-operation.js]: I added this, however it directly conflicts with our current lint .js configuration -  https://eslint.org/docs/latest/rules/dot-notation - we can have a conversation re: this/moving forward, but I think the resolution likely falls outside the ticket scope.

File: packages/api/endpoints/granules.js
Comment (marchuffnagle) [reply packages/api/endpoints/granules.js:585]: Sounds like a good architecture meeting topic.

**Comment by Jkovarik** [comment]
@marchuffnagle  I believe the commentary has been addressed, thanks for the feedback 💯    I was forced to move zod-utils out of it's separate package for npm reasons, we should chat about that, but I believe this is good for a re-review.

File: packages/api/lambdas/bulk-operation.js
Comment (marchuffnagle) [reply packages/api/lambdas/bulk-operation.js:90]: Yeah, that conflict is resolved in this PR - https://github.com/nasa/cumulus/pull/3239

**Review by marchuffnagle — APPROVED** [review]
Approved, assuming there are no objections to these changes - 6a80e45f3249e0079b28630eb3e4bc13e53df689

File: packages/db/src/lib/granule.ts
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:22]: So, I grok restricting the typing, in that our callers currently aren't using a transaction, however most of the db/lib functions are intentionally inclusive of transactions.         This is likely a user-facing breaking change,   we should chat briefly about it.

Comment (marchuffnagle) [reply packages/db/src/lib/granule.ts:22]: The `Transaction` interface extends the `Knex` interface, so any `Knex.Transaction` will meet this requirement.

Comment (marchuffnagle) [reply packages/db/src/lib/granule.ts:22]: Or, said another way, if a user passes a `Knex.Transaction` in here, it'll work just fine because a `Transaction` is a valid `Knex`.

File: packages/db/src/lib/granule.ts
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:23]: While I don't functionally disagree with constraining the type in *this* case, the proposed change is again a breaking change from what was in the file prior.   

 We should consider this in main, but this changeset is being proposed against the 11.1.x baseline and given it's a public API I think we should revert the typing update here.   Additionally it's inconsistent with our current usage of it in places like `granule-remove-from-cmr.js`.... which raises the question of why the typedef/import is passing the check at all 👀

File: packages/db/src/lib/granule.ts
Comment (Jkovarik) [reply packages/db/src/lib/granule.ts:22]: Ah - so it does.  I'm wondering if we missed that in an upgrade typings-wise, or if it was just a mistake.

**Comment by Jkovarik** [comment]
Thanks for the reviews @marchuffnagle  @npauzenga !

**Pull request merged**
